tools/xdp/s2n-quic-xdp/src/syscall.rs

```diff
@@ -0,0 +1,342 @@
+    const SO_NETNS_COOKIE: libc::c_int = 71;
+
+    let mut cookie = 0u64;
+    let mut optlen = size_of::<u64>() as libc::socklen_t;
+
+    let ret = unsafe {
+        libc::getsockopt(
+            fd.as_raw_fd(),
+            libc::SOL_SOCKET,
+            SO_NETNS_COOKIE,
+            &mut cookie as *mut _ as _,
+            &mut optlen,
+        )
+    };
+
+    if ret == 0 {
+        return Ok(cookie);
+    }
+
+    // the cookie syscall is not supported here so return the default value of 1
+    // https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L55
+    if errno::errno().0 == libc::ENOPROTOOPT {
+        return Ok(1);
+    }
+
+    Err(io::Error::last_os_error())
+}
+
+/// Configures the fill ring size of the provided AF_XDP socket
+///
+/// See [xsk.c](https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L257).
+#[inline]
+pub fn set_fill_ring_size<Fd: AsRawFd>(fd: &Fd, len: u32) -> Result<()> {
+    set_xdp_option(fd, SocketOptions::UmemFillRing, &len)
+}
+
+/// Configures the completion ring size of the provided AF_XDP socket
+///
+/// See [xsk.c](https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L263).
+#[inline]
+pub fn set_completion_ring_size<Fd: AsRawFd>(fd: &Fd, len: u32) -> Result<()> {
+    set_xdp_option(fd, SocketOptions::UmemCompletionRing, &len)
+}
+
+/// Configures the RX ring size of the provided AF_XDP socket
+///
+/// See [xsk.c](https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L1081).
+#[inline]
+pub fn set_rx_ring_size<Fd: AsRawFd>(fd: &Fd, len: u32) -> Result<()> {
+    set_xdp_option(fd, SocketOptions::RxRing, &len)
+}
+
+/// Configures the TX ring size of the provided AF_XDP socket
+///
+/// See [xsk.c](https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L1093).
+#[inline]
+pub fn set_tx_ring_size<Fd: AsRawFd>(fd: &Fd, len: u32) -> Result<()> {
+    set_xdp_option(fd, SocketOptions::TxRing, &len)
+}
+
+/// Configures the UMEM options for the provided AF_XDP socket
+///
+/// See [xsk.c](https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L344).
+#[inline]
+pub fn set_umem<Fd: AsRawFd>(fd: &Fd, reg: &UmemReg) -> Result<()> {
+    set_xdp_option(fd, SocketOptions::UmemReg, reg)
+}
+
+/// Binds the provided AF_XDP socket to an address
+#[inline]
+pub fn bind<Fd: AsRawFd>(fd: &Fd, addr: &mut Address) -> Result<()> {
+    libc!(bind(
+        fd.as_raw_fd(),
+        addr as *mut _ as _,
+        size_of::<Address>() as _
+    ))?;
+    Ok(())
+}
+
+/// Notifies the kernel to send any packets on the TX ring
+///
+/// This should be called after checking if the TX ring needs a wake up.
```

Comment:
what does this mean and how do you check? what happens if this is not done? should this fn be `unsafe`?

Reply:
It's checked via the ring flags. If it doesn't get called then the kernel doesn't wake up the driver to consume the entries. I don't think it should be unsafe since this function isn't actually what enforces this. Later on there will be safe abstractions that enforce the wake up call when the ring is modified.